docs/pages/kubernetes-access/controls.mdx

```diff
@@ -639,16 +639,25 @@ these values to a list of RBAC principals it will use to write impersonation
 headers later on.
 
 Next, the Kubernetes Service checks each of the user's roles for `spec.deny`
-conditions. If one role's `spec.deny.kubernetes_labels` or
-`spec.deny.kubernetes_resources` fields match the user's request, the Kubernetes
-Service looks up the role's `spec.deny.kubernetes_groups` and
-`spec.deny.kubernetes_users` fields. It removes each of these values from the
-list of users and groups it created earlier, denying the user access to these
-RBAC principals.
+conditions that might apply to the request:
+
+- For `spec.deny.kubernetes_labels`, if any labels match the cluster's labels 
+  for the request, the Service denies the request before it is made to the 
+  Kubernetes API. 
+- For `spec.deny.kubernetes_resources`, if the resource matches a `deny` condition, 
+  the associated `deny.kubernetes_groups` and `deny.kubernetes_users` are 
+  removed from the previously developed list of allowed RBAC principals from 
+  `spec.allow`. 
+
+Please note that a match in `spec.deny.kubernetes_labels` leads to an immediate 
+denial without a chance to adjust the user's impersonation data based on allow 
+rules of other roles they may have. On the other hand, a match in 
+`spec.deny.kubernetes_resources` only affects the user's RBAC principals but does 
+not result in an immediate denial.
```

Comment:
Same as before. 
If a role specified `deny.kubernetes_groups=system:masters` and you have other role that grants `allow.kubernetes_groups=["system:masters", "view"]` the following happens:
- kubectl get pods -> uses `view` group. We collect `system:masters` and `view` from the allow clause, and after that we exclude the `system:masters`
- kubectl get pods --as user --as-group= view -> uses `view` group. We collect `system:masters` and `view` from the allow clause, and after that we exclude the `system:masters`. We compare against the request headers and the request goes through
- kubectl get pods --as user --as-group=system:masters -> request is denied. The user can't use the system:masters group
- kubectl get pods --as user --as-group=system:masters --as-group=view -> request is denied. The user can't use the system:masters group